frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,135 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+export function Tree({ jsonData, onNodeClick }) {
+  const svgRef = useRef();
+  const wrapperRef = useRef();
+  const [dimensions, setDimensions] = useState({ width: 0, height: 0 });
+
+  // We save data to see if it changed
+  const previouslyRenderedData = usePrevious(jsonData);
+
+  // This effect updates the dimensions state every time the user
+  // resizes their window
+  useEffect(() => {
+    const observeTarget = wrapperRef.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions({
+          // add margins to prevent chopped off content
+          width: entry.contentRect.width - 200,
+          height: entry.contentRect.height - 50,
+        });
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [wrapperRef]);
+
+  // Will be called initially and on every data change
+  useEffect(() => {
+    const svg = select(svgRef.current);
+    const { width, height } = dimensions;
+    const marginLeft = 70;
+    const marginTop = 30;
+
+    // Transform hierarchical data
+    const root = hierarchy(jsonData);
+    const treeLayout = tree().size([height, width]);
+
+    // Creates the links between nodes
+    const linkGenerator = linkHorizontal()
+      .x((link) => link.y + marginLeft)
+      .y((link) => link.x + marginTop);
+
+    // Enrich hierarchical data with coordinates
+    treeLayout(root);
+
+    // Create the node group, which will hold the nodes and labels
+    const nodeGroup = svg.selectAll(".node-group").data(root.descendants());
+    const nodeGroupEnter = nodeGroup.enter().append("g");
```

Comment:
what are the intuitive uses of `nodeGroupEnter` and `enteringAndUpdatingNodes`? Are there any better names for these?